cmd/ipfs/daemon.go

```diff
@@ -806,6 +806,12 @@ func serveHTTPGateway(req *cmds.Request, cctx *oldcmds.Context) (<-chan error, e
 		return nil, fmt.Errorf("serveHTTPGateway: ConstructNode() failed: %s", err)
 	}
 
+	if len(listeners) > 0 {
+		if err := node.Repo.SetGatewayAddr(listeners[0].Addr()); err != nil {
```

Comment:
It seems quite arbitrary to pick the index `0`, I would maybe add a filter to pick `localhost` if possible but I don't think it's worth fixing.